Rounding mode support for hexadecimal float parsing, with improvements to error handling

quaternic opened the pull request:
Parsing errors are now bubbled up part of the way, but that needs some more work.

Rounding should be correct, and the `Status` returned by `parse_any` should have the correct bits set. These are used for the current (unchanged) behaviour of the surface level functions like `hf64`: panic on invalid inputs, or values that aren't exactly representable.

tgross35 commented on src/math/support/hex_float.rs:
```diff
@@ -2,149 +2,258 @@
 
-    // finally, set the sign bit if necessary
-    sig | ((neg as u128) << (bits - 1))
+/// Divide by 4, rounding accor
```
Did the rest of this get cut off?

quaternic replied:
Yeah. The cutting off happened when I couldn't think of a good wording while writing it and moved to something else and didn't notice it later.
```suggestion
/// Divide by 4, rounding with the given mode
```

tgross35 reviewed (approved):
Sorry I disappeared for a bit soon after you posted this, but it looks great! Just one possible doc fix.

quaternic commented:
@tgross35 
> Sorry I disappeared for a bit soon after you posted this, but it looks great! Just one possible doc fix.
Thanks! I'm sorry for taking my time as well. I fixed the comment and rebased; and CI seems to still be happy with it. Could you squash this while merging?

CrazyboyQCD commented:
There are some magic numbers and literals, it would be good to make them constants for better readability.

quaternic commented on src/math/support/hex_float.rs:
```diff
@@ -2,149 +2,258 @@
+    let t = (x as u32) & 0b111;
+    x >>= 2;
+    match round {
+        Round::Nearest => x + ((0b11001000_u8 >> t) & 1) as u128,
```
@CrazyboyQCD If you meant this, yeah, it could use an explanatory comment. I don't think extracting to a constant would help since it only makes sense in the context it is used in. ( Though this and the previous function could be more generally useful as utilities for `Round`. )
```suggestion
        // Look-up-table on the last three bits for when to round up 
        Round::Nearest => x + ((0b11001000_u8 >> t) & 1) as u128,
        
```
Or was there something else?

tgross35 replied:
I just applied this since it's the only case I see, @CrazyboyQCD feel free to submit a followup PR or ask for clarification if there is anything else.

tgross35 commented on src/math/support/hex_float.rs:
```diff
@@ -2,149 +2,260 @@
+    let t = (x as u32) & 0b111;
+    x >>= 2;
+    match round {
+        // Look-up-table on the last three bits for when to round up 
+        Round::Nearest => x + ((0b11001000_u8 >> t) & 1) as u128,
+        
+        Round::Negative => x,
```
Rustfmt

```suggestion
        // Look-up-table on the last three bits for when to round up
        Round::Nearest => x + ((0b11001000_u8 >> t) & 1) as u128,

        Round::Negative => x,
```
